core/controllers/collection_viewer.py

```diff
@@ -0,0 +1,133 @@
+    """Page describing a single collection."""
+
+    PAGE_NAME_FOR_CSRF = 'collection'
+
+    @require_collection_playable
+    def get(self, collection_id):
+        """Handles GET requests."""
+        try:
+            collection = collection_services.get_collection_by_id(
+                collection_id)
+        except Exception as e:
+            raise self.PageNotFoundException(e)
+
+        version = collection.version
+
+        if not rights_manager.Actor(self.user_id).can_view(
+                rights_manager.ACTIVITY_TYPE_COLLECTION, collection_id):
+            raise self.PageNotFoundException
+
+        self.values.update({
+            'can_edit': (
+                bool(self.username) and
+                self.username not in config_domain.BANNED_USERNAMES.value and
+                rights_manager.Actor(self.user_id).can_edit(
+                    rights_manager.ACTIVITY_TYPE_COLLECTION, collection_id)
+            ),
+            'collection_id': collection_id,
+            'collection_title': collection.title,
+            'is_private': rights_manager.is_collection_private(collection_id),
+            'meta_name': collection.title,
+            'meta_description': utils.make_first_letter_uppercase(
+                collection.objective)
+        })
+
+        self.render_template('collection_player/collection_player.html')
+
+
+class CollectionDataHandler(base.BaseHandler):
+    """Provides the data for a single collection."""
+
+    def get(self, collection_id):
+        """Populates the data on the individual collection page."""
+        try:
+            collection = collection_services.get_collection_by_id(
+                collection_id)
+        except Exception as e:
+            raise self.PageNotFoundException(e)
+
+        exp_ids = [
+            collection_node.exploration_id
+            for collection_node in collection.nodes]
+
+        exp_summaries = (
+            exp_services.get_exploration_summaries_matching_ids(exp_ids))
+
+        exp_titles_dict = {}
+        for (ind, exp_id) in enumerate(exp_ids):
+            exp_summary = exp_summaries[ind]
+            exp_titles_dict[exp_id] = exp_summary.title if exp_summary else ''
+
+        # TODO(bhenning): Users should not be recommended explorations they
+        # have completed outside the context of a collection.
+        next_exploration_ids = None
+        if self.user_id:
+            completed_exploration_ids = (
+                collection_services.get_completed_exploration_ids(
+                    self.user_id, collection_id))
+            next_exploration_ids = collection.get_next_exploration_ids(
+                completed_exploration_ids)
+        else:
+            # If the user is not logged in or they have not completed any of the
+            # explorations yet within the context of this collection, recommend the
+            # initial explorations.
+            next_exploration_ids = collection.init_exploration_ids
```

Comment:
It strikes me that integration tests for both the logged-in and logged-out user experiences are quite important.